library/alloc/src/vec/source_iter_marker.rs

```diff
@@ -1,3 +1,100 @@
+///! Inplace iterate-and-collect specialization for `Vec`
+///!
+///! This applies to iterators in the shape of
+///! `source.adapter().adapter().adapter().collect::<Vec<U>>()`
+///! where `source` is a `Vec`, `Box<[]>` or `BinaryHeap`, the adapters each
```

Comment:
This is a bit unclear, can `.adapter()` be `.into_iter()`? If not then `source` should be the corresponding `IntoIterator::IntoIter` type of those listed.

`Box<[]>` doesn't implement `IntoIterator` (I feel like this is something that could have be done with the 2021 edition). You can still get an owned iterator over its contents but you first have to convert it to a `Vec` with `From`/`Into`, so it falls under the `Vec` case.

Reply:
> This is a bit unclear, can .adapter() be .into_iter()? If not then source should be the corresponding IntoIterator::IntoIter type of those listed.

I'll add the `into_iter()`.

>  You can still get an owned iterator over its contents but you first have to convert it to a Vec with From/Into, so it falls under the Vec case.

Indeed, the key part is that the conversion doesn't perform any allocations either so vec's iter basically is box's iter.

Reply:
I went with calling `source` the respective `IntoIter`s since that'll be what holds the allocation which the rest of the text talks about. The original collection will have ceased to exist at that point.